File: iosApp/iosApp/Modules/NFTSend/NFTSendPresenter.swift
// Created by web3d4v on 04/08/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import web3walletcore

enum NFTSendPresenterEvent {
    case dismiss
    case addressChanged(to: String)
    case pasteAddress
    case saveAddress
    case feeChanged(to: NetworkFee)
    case qrCodeScan
    case feeTapped
    case review
}

protocol NFTSendPresenter: AnyObject {
    func present()
    func handle(_ event: NFTSendPresenterEvent)
}

final class DefaultNFTSendPresenter {
    private weak var view: NFTSendView?
    private let wireframe: NFTSendWireframe
    private let interactor: NFTSendInteractor
    private let context: NFTSendWireframeContext
    
    private var sendTapped = false
    private var address: String?
    private var fee: NetworkFee?

    private var items = [NFTSendViewModel.Item]()
    private var fees = [NetworkFee]()

    init(
        view: NFTSendView,
        wireframe: NFTSendWireframe,
        interactor: NFTSendInteractor,
        context: NFTSendWireframeContext
    ) {
        self.view = view
        self.wireframe = wireframe
        self.interactor = interactor
        self.context = context
        refreshFees()
    }
}

extension DefaultNFTSendPresenter: NFTSendPresenter {

    func present() {
        updateView(
            with: [
                .nft(context.nftItem),
                .address(
                    .init(
                        placeholder: Localized("networkAddressPicker.to.address.placeholder", context.network.name),
                        value: nil,
                        isValid: false,
                        becomeFirstResponder: true
                    )
                ),
                .send(
                    .init(
                        tokenNetworkFeeViewModel: .init(
                            estimatedFee: estimatedFee(),
                            feeName: feeName()
                        ),
                        buttonState: .ready
                    )
                )
            ]
        )
    }

    func handle(_ event: NFTSendPresenterEvent) {
        switch event {
        case .dismiss:
            wireframe.dismiss()
        case .qrCodeScan:
            view?.dismissKeyboard()
            wireframe.navigate(
                to: .qrCodeScan(network: context.network, onCompletion: onQRCodeScanned())
            )
        case .saveAddress:
            wireframe.navigate(to: .underConstructionAlert)
        case let .addressChanged(address):
            if
                let currentAddress = self.address,
                let formattedAddress = formattedAddress,
                formattedAddress.hasPrefix(address),
                formattedAddress.count == (address.count + 1)
            {
                updateAddress(with: String(currentAddress.prefix(currentAddress.count - 1)))
            } else {
                updateView(address: address)
            }
        case .pasteAddress:
            let clipboard = UIPasteboard.general.string ?? ""
            let isValid = context.network.isValidAddress(input: clipboard)
            guard isValid else { return }
            updateView(address: clipboard)
        case let .feeChanged(fee):
            self.fee = fee
            updateCTA()
        case .feeTapped:
            view?.presentFeePicker(with: fees)
        case .review:
            sendTapped = true
            let isValidAddress = context.network.isValidAddress(input: address ?? "")
            guard let address = address, isValidAddress  else {
                updateView(shouldAddressBecomeFirstResponder: true)
                return
            }
            guard let walletAddress = interactor.walletAddress, let fee = fee else { return }
            let context = ConfirmationWireframeContext.SendNFT(
                data: .init(
                    network: context.network,
                    addressFrom: walletAddress,
                    addressTo: address,
                    nftItem: context.nftItem,
                    networkFee: fee
                )
            )
            wireframe.navigate(to: .confirmSendNFT(context: context))
        }
    }
}

private extension DefaultNFTSendPresenter {
    
    func refreshFees() {
        fees = interactor.networkFees(network: context.network)
        guard fee == nil, !fees.isEmpty else { return }
        fee = fees[0]
    }
    
    func updateView(with items: [NFTSendViewModel.Item]) {
        view?.update(
            with: .init(
                title: Localized("nftSend.title"),
                items: items
            )
        )
    }
    
    func onQRCodeScanned() -> (String) -> Void {
        { [weak self] address in self?.updateAddress(with: address) }
    }
    
    func updateView(
        address: String? = nil,
        shouldAddressBecomeFirstResponder: Bool = false,
        amount: Double? = nil
    ) {
        updateAddress(
            with: address ?? self.address ?? "",
            becomeFirstResponder: shouldAddressBecomeFirstResponder
        )
        updateCTA()
    }
    
    func updateAddress(
        with address: String,
        becomeFirstResponder: Bool = false
    ) {
        if !address.contains("...") {
            self.address = address
        }
        let isValid = context.network.isValidAddress(input: self.address ?? "")
        updateView(
            with: [
                .address(
                    .init(
                        placeholder: Localized("networkAddressPicker.to.address.placeholder", context.network.name),
                        value: formattedAddress ?? address,
                        isValid: isValid,
                        becomeFirstResponder: becomeFirstResponder
                    )
                )
            ]
        )
    }
    
    var formattedAddress: String? {
        guard let address = address else { return nil }
        guard context.network.isValidAddress(input: address) else { return nil }
        return Formatters.Companion.shared.networkAddress.format(
            address: address,
            digits: 8,
            network: context.network
        )
    }

    func updateCTA() {
        let isValidAddress = context.network.isValidAddress(input: address ?? "")
        let buttonState: NFTSendViewModel.Send.State
        if !sendTapped { buttonState = .ready }
        else if !isValidAddress { buttonState = .invalidDestination }
        else { buttonState = .ready }
        updateView(
            with: [
                .send(
                    .init(
                        tokenNetworkFeeViewModel: .init(
                            estimatedFee: estimatedFee(),
                            feeName: feeName()
                        ),
                        buttonState: buttonState
                    )
                )
            ]
        )
    }
    
    func estimatedFee() -> [Formatters.Output] {
        guard let fee = fee else { return [Formatters.OutputNormal(value: "-")] }
        var outputFormat = Formatters.Companion.shared.currency.format(
            amount: fee.amount, currency: fee.currency, style: Formatters.StyleCustom(maxLength: 10)
        )
        let min: Double = Double(fee.seconds) / Double(60)
        var value = ""
        if min > 1 {
            value = " ~ \(min.toString(decimals: 0)) \(Localized("min"))"
        } else {
            value = " ~ \(fee.seconds) \(Localized("sec"))"
        }
        outputFormat.append(
            Formatters.OutputNormal(value: value)
        )
        return outputFormat
    }
    
    func feeName() -> String {
        guard let fee = fee else { return "-" }
        return fee.name
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/nftSend/NFTSendPresenter.kt
package com.sonsofcrypto.web3walletcore.modules.nftSend

import com.sonsofcrypto.web3lib.formatters.Formatters
import com.sonsofcrypto.web3lib.types.NetworkFee
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3walletcore.common.viewModels.NetworkAddressPickerViewModel
import com.sonsofcrypto.web3walletcore.common.viewModels.NetworkFeeViewModel
import com.sonsofcrypto.web3walletcore.extensions.Localized
import com.sonsofcrypto.web3walletcore.extensions.toNetworkFeeViewModel
import com.sonsofcrypto.web3walletcore.modules.confirmation.ConfirmationWireframeContext
import com.sonsofcrypto.web3walletcore.modules.nftSend.NFTSendViewModel.ButtonState.INVALID_DESTINATION
import com.sonsofcrypto.web3walletcore.modules.nftSend.NFTSendViewModel.ButtonState.READY
import com.sonsofcrypto.web3walletcore.modules.nftSend.NFTSendWireframeDestination.ConfirmSendNFT
import com.sonsofcrypto.web3walletcore.modules.nftSend.NFTSendWireframeDestination.Dismiss
import com.sonsofcrypto.web3walletcore.modules.nftSend.NFTSendWireframeDestination.QRCodeScan
import com.sonsofcrypto.web3walletcore.modules.nftSend.NFTSendWireframeDestination.UnderConstructionAlert

sealed class NFTSendPresenterEvent {
    object QrCodeScan: NFTSendPresenterEvent()
    data class AddressChanged(val value: String): NFTSendPresenterEvent()
    data class PasteAddress(val value: String): NFTSendPresenterEvent()
    object SaveAddress: NFTSendPresenterEvent()
    object NetworkFeeTapped: NFTSendPresenterEvent()
    data class NetworkFeeChanged(val value: NetworkFee): NFTSendPresenterEvent()
    object Review: NFTSendPresenterEvent()
    object Dismiss: NFTSendPresenterEvent()
}

interface NFTSendPresenter {
    fun present()
    fun handle(event: NFTSendPresenterEvent)
}

class DefaultNFTSendPresenter(
    private val view: WeakRef<NFTSendView>,
    private val wireframe: NFTSendWireframe,
    private val interactor: NFTSendInteractor,
    private val context: NFTSendWireframeContext,
): NFTSendPresenter {
    private var address: String? = null
    private var networkFees: List<NetworkFee> = interactor.networkFees(context.network)
    private var networkFee: NetworkFee? = networkFees.firstOrNull()
    private var sendTapped = false

    override fun present() { updateView(addressBecomeFirstResponder = true) }

    override fun handle(event: NFTSendPresenterEvent) {
        when (event) {
            is NFTSendPresenterEvent.QrCodeScan -> {
                view.get()?.dismissKeyboard()
                wireframe.navigate(QRCodeScan { address = it; updateView() })
            }
            is NFTSendPresenterEvent.AddressChanged -> {
                updateAddressIfNeeded(event.value)
                updateView()
            }
            is NFTSendPresenterEvent.PasteAddress -> {
                val isValidAddress = context.network.isValidAddress(event.value)
                if (!isValidAddress) return
                address = event.value
                updateView()
            }
            is NFTSendPresenterEvent.SaveAddress -> wireframe.navigate(UnderConstructionAlert)
            is NFTSendPresenterEvent.NetworkFeeTapped -> {
                view.get()?.presentNetworkFeePicker(networkFees)
            }
            is NFTSendPresenterEvent.NetworkFeeChanged -> {
                networkFee = event.value
                updateView()
            }
            is NFTSendPresenterEvent.Review -> {
                sendTapped = true
                val context = confirmationWireframeSendNFTContext() ?: return
                wireframe.navigate(ConfirmSendNFT(context))
            }
            is NFTSendPresenterEvent.Dismiss -> wireframe.navigate(Dismiss)
        }
    }

    private fun updateView(addressBecomeFirstResponder: Boolean = false) =
        view.get()?.update(viewModel(addressBecomeFirstResponder))

    private fun viewModel(addressBecomeFirstResponder: Boolean = false): NFTSendViewModel =
        NFTSendViewModel(
            Localized("nftSend.title"),
            listOf(
                nftViewModel(),
                networkAddressViewModel(addressBecomeFirstResponder),
                sendViewModel(),
            )
        )

    private fun nftViewModel(): NFTSendViewModel.Item.Nft =
        NFTSendViewModel.Item.Nft(context.nftItem)

    private fun networkAddressViewModel(
        becomeFirstResponder: Boolean = false,
    ): NFTSendViewModel.Item.Address {
        val data = NetworkAddressPickerViewModel(
            Localized("networkAddressPicker.to.address.placeholder", context.network.name),
            formattedAddress,
            context.network.isValidAddress(address ?: ""),
            becomeFirstResponder
        )
        return NFTSendViewModel.Item.Address(data)
    }

    private val formattedAddress: String get() = formattedAddress(address)

    private fun formattedAddress(address: String?): String {
        address ?: return ""
        if (!context.network.isValidAddress(address)) { return address }
        return Formatters.networkAddress.format(address, 8, context.network)
    }

    private fun sendViewModel(): NFTSendViewModel.Item.Send =
        NFTSendViewModel.Item.Send(
            networkFeeViewModel(),
            buttonStateViewModel()
        )

    private fun networkFeeViewModel(): NetworkFeeViewModel {
        val networkFee = networkFee ?: return emptyNetworkFeeViewModel()
        val mul = interactor.fiatPrice(networkFee.currency)
        return networkFee.toNetworkFeeViewModel(mul)
    }

    private fun emptyNetworkFeeViewModel(): NetworkFeeViewModel =
        NetworkFeeViewModel("-", listOf(), listOf(), listOf())

    private fun buttonStateViewModel(): NFTSendViewModel.ButtonState =
        if (!(context.network.isValidAddress(address ?: ""))) { INVALID_DESTINATION }
        else { READY }

    private fun updateAddressIfNeeded(value: String) =
        if (value.isEmpty()) {
            address = value
        } else if (formattedAddress.startsWith(value) &&
            formattedAddress.count() == value.length + 1) {
            address = address?.dropLast(1)
        } else if (
            formattedAddress != formattedAddress(value) &&
            (address?.length ?: 0) < value.length
        ) {
            address = value
        } else {
            // nothing
        }

    private fun confirmationWireframeSendNFTContext(): ConfirmationWireframeContext.SendNFT? {
        val addressFrom = interactor.walletAddress ?: return null
        val addressTo = address ?: return null
        val networkFee = networkFee ?: return null
        val data = ConfirmationWireframeContext.SendNFTContext(
            context.network,
            addressFrom,
            addressTo,
            context.nftItem,
            networkFee,
        )
        return ConfirmationWireframeContext.SendNFT(data)
    }
}
